app/models/ems_refresh/save_inventory.rb

```diff
@@ -246,12 +246,39 @@ def save_networks_inventory(hardware, hashes, mode = :refresh)
   def save_system_services_inventory(parent, hashes, mode = :refresh)
     return if hashes.nil?
 
-    deletes = case mode
-              when :refresh then nil
-              when :scan    then :use_association
-              end
+    #######
+    # tripleo specific
+    #######
+
+    # if parent is OpenStack Cloud
+    # and if parent have Infra provider
+    if parent.kind_of?(ManageIQ::Providers::Openstack::CloudManager)
+      infra_ems = parent.provider \
+        && parent.provider.kind_of?(ManageIQ::Providers::Openstack::Provider) \
+        && parent.provider.infra_ems
+      if infra_ems
+        # for each host
+        infra_ems.hosts.map do |host|
+          # select hashes with that hostname
+          hashes_for_host = hashes.select do |hash|
+            hash[:host] == host.hypervisor_hostname
+            # and put host instead of hostname there
+          end.map do |hash|
+            hash[:host] = host
+            hash
+          end
+          # save system_services for one host
+          save_inventory_multi(host.system_services, hashes_for_host, [], [:name])
```

Comment:
Hm, it's bad it's so hard to access the one table from multiple places. Maybe we should bring new table cloud_services, which would optionally belong to host and system_service? What do you think? It could clean the code a bit.